### PR TITLE
[BE] bump `optree` version to 0.12.1

### PR DESCRIPTION
**Pull request by XuehaiPan**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #130140
* #130141
* #130144
* __->__ #130139
* #130895

0.12.0 Major Updates:

- Add context manager to temporarily set the dictionary sorting mode
- Add accessor APIs
- Use `stable` tag for `pybind11` for Python 3.13 support
- Fix potential segmentation fault for pickling support

0.12.1 Updates:

- Fix warning regression during import when launch with strict warning filters

Closes #130155

### CODE REVIEW

**Comment by clee2000** [comment]
Please also pin optree in windows CI https://github.com/pytorch/pytorch/blob/a33ee73a284977b7956b43c069fb238f4784a7bb/.github/workflows/_win-test.yml#L191 

The update to 0.12.0 got picked up and is currently failing.  I have https://github.com/pytorch/pytorch/pull/130155 to pin it to 0.11.0 to restore the current CI

File: torch/utils/_cxx_pytree.py [hunk 1]
```diff
@@ -273,7 +273,7 @@ def tree_flatten(
     >>> from collections import OrderedDict
     >>> tree = OrderedDict([('b', (2, [3, 4])), ('a', 1), ('c', None), ('d', 5)])
     >>> tree_flatten(tree)
-    ([2, 3, 4, 1, None, 5], PyTreeSpec(OrderedDict([('b', (*, [*, *])), ('a', *), ('c', *), ('d', *)]), NoneIsLeaf))
+    ([2, 3, 4, 1, None, 5], PyTreeSpec(OrderedDict({'b': (*, [*, *]), 'a': *, 'c': *, 'd': *}), NoneIsLeaf))
```
Comment (XuehaiPan) [inline torch/utils/_cxx_pytree.py:276]: > The update to 0.12.0 got picked up and is currently failing.

@clee2000 The doctest fix is included.

**Review by zou3519 — APPROVED** [review]

**Comment by XuehaiPan** [comment]
I can confirm the test failure is caused by `sympy` (ref #130710 and #130729). Pinning `sym < 1.13` resolves the unrelated failures.

- #130710
- #130729

**Comment by albanD** [comment]
Is this new pytree version working with 3.13t ?

**Comment by XuehaiPan** [comment]
> Is this new pytree version working with 3.13t ?

Yes, it is tested against Python 3.13t.

File: test/functorch/test_eager_transforms.py [hunk 2]
```diff
@@ -3404,7 +3404,7 @@ def f(x):
     @onlyCPU
     def test_no_warning_on_import_functorch(self, device):
         out = subprocess.check_output(
-            [sys.executable, "-W", "all", "-c", "import functorch"],
+            [sys.executable, "-W", "always", "-c", "import functorch"],
```
Comment (albanD) [inline test/functorch/test_eager_transforms.py:3407]: I guess this is just an unrelated bugfix?

Comment (XuehaiPan) [reply test/functorch/test_eager_transforms.py:3407]: `-Wall` is aliased to `-Walways` ([Python docs](https://docs.python.org/3/using/cmdline.html#cmdoption-W)). They are the same.

Comment (albanD) [reply test/functorch/test_eager_transforms.py:3407]: Why this change then?

**Pull request closed without merging**